python/ray/data/dataset.py

```diff
@@ -3988,12 +3989,19 @@ def materialize(self) -> "MaterializedDataset":
             )
             for block_with_metadata in blocks_with_metadata
         ]
-
-        # Create a new logical plan whose input is the existing data
-        # from the the old Dataset.
-        copy._logical_plan = LogicalPlan(InputData(input_data=ref_bundles))
-
-        return copy
+        logical_plan = LogicalPlan(InputData(input_data=ref_bundles))
+        output = MaterializedDataset(
+            ExecutionPlan(
+                blocks,
+                DatasetStats(stages={"FromMaterializedBlocks": metadata}, parent=None),
+                run_by_consumer=False,
+            ),
+            0,
+            True,
+            logical_plan,
+        )
```

Comment:
could you explain why this fixes the issues?

Reply:
This is mirroring the logical plan change, to make sure the materialized dataset is decoupled from the previous dataset plan. This avoids needing to check correctly whether the plan has snapshotted outputs or not, which had bugs for the streaming split case.

Reply:
Thanks. Then I think the `copy` variable is not needed anymore? we can directly call `self._plan.execute(force_read=True)`?

Reply:
We still need the copy to avoid changing the materialization state of the base dataset (i.e., only the return of materialize() should be fully executed).

Reply:
sounds good.